# 2022-12-04 Data Science Homework 5 Problem 1

###### HsuanYu CHOU

In [1]:
import numpy as np

In [2]:
def gini(n):
    s = sum(n)
    return 1 - (n[0] / s) ** 2 - (n[1] / s) ** 2

def wgini(n):
    return sum(n) * gini(n)

In [3]:
data = []

with open('./data.tsv') as f:
    lines = f.readlines()
for line in lines:
    l = line[:-1].split('\t')
    data.append(l)

data = np.array(data)
print(data)

[['Customer ID' 'Gender' 'Car Type' 'Shirt Size' 'Class']
 ['1' 'M' 'Family' 'Small' 'C0']
 ['2' 'M' 'Family' 'Medium' 'C0']
 ['3' 'M' 'Family' 'Extra Large' 'C0']
 ['4' 'M' 'Sports' 'Extra Large' 'C0']
 ['5' 'M' 'Family' 'Large' 'C0']
 ['6' 'M' 'Family' 'Extra Large' 'C0']
 ['7' 'M' 'Luxury' 'Extra Large' 'C0']
 ['8' 'F' 'Family' 'Small' 'C0']
 ['9' 'F' 'Family' 'Medium' 'C0']
 ['10' 'F' 'Luxury' 'Large' 'C0']
 ['11' 'F' 'Luxury' 'Large' 'C0']
 ['12' 'M' 'Sports' 'Medium' 'C1']
 ['13' 'M' 'Sports' 'Large' 'C1']
 ['14' 'M' 'Sports' 'Medium' 'C1']
 ['15' 'F' 'Sports' 'Small' 'C1']
 ['16' 'F' 'Luxury' 'Small' 'C1']
 ['17' 'F' 'Luxury' 'Small' 'C1']
 ['18' 'F' 'Sports' 'Medium' 'C1']
 ['19' 'F' 'Luxury' 'Medium' 'C1']
 ['20' 'F' 'Luxury' 'Medium' 'C1']]


## ROOT

In [4]:
root = data[1:]
print(root)

[['1' 'M' 'Family' 'Small' 'C0']
 ['2' 'M' 'Family' 'Medium' 'C0']
 ['3' 'M' 'Family' 'Extra Large' 'C0']
 ['4' 'M' 'Sports' 'Extra Large' 'C0']
 ['5' 'M' 'Family' 'Large' 'C0']
 ['6' 'M' 'Family' 'Extra Large' 'C0']
 ['7' 'M' 'Luxury' 'Extra Large' 'C0']
 ['8' 'F' 'Family' 'Small' 'C0']
 ['9' 'F' 'Family' 'Medium' 'C0']
 ['10' 'F' 'Luxury' 'Large' 'C0']
 ['11' 'F' 'Luxury' 'Large' 'C0']
 ['12' 'M' 'Sports' 'Medium' 'C1']
 ['13' 'M' 'Sports' 'Large' 'C1']
 ['14' 'M' 'Sports' 'Medium' 'C1']
 ['15' 'F' 'Sports' 'Small' 'C1']
 ['16' 'F' 'Luxury' 'Small' 'C1']
 ['17' 'F' 'Luxury' 'Small' 'C1']
 ['18' 'F' 'Sports' 'Medium' 'C1']
 ['19' 'F' 'Luxury' 'Medium' 'C1']
 ['20' 'F' 'Luxury' 'Medium' 'C1']]


In [5]:
n_01 = np.array([0, 0])
n_M, n_F = np.array([0, 0]), np.array([0, 0])
n_Fa, n_Sp, n_Lu = np.array([0, 0]), np.array([0, 0]), np.array([0, 0])
n_Sm, n_Me, n_La, n_Ex = np.array([0, 0]), np.array([0, 0]), np.array([0, 0]), np.array([0, 0])

for i in root[:, 0]:
    row = data[int(i)]

    n_01[1 - int(row[-1] == 'C0')] += 1

    if row[1] == 'M':
        n_M[1 - int(row[-1] == 'C0')] += 1
    if row[1] == 'F':
        n_F[1 - int(row[-1] == 'C0')] += 1

    if row[2] == 'Family':
        n_Fa[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Sports':
        n_Sp[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Luxury':
        n_Lu[1 - int(row[-1] == 'C0')] += 1

    if row[3] == 'Small':
        n_Sm[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Medium':
        n_Me[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Large':
        n_La[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Extra Large':
        n_Ex[1 - int(row[-1] == 'C0')] += 1

print(n_01)
print(n_M, n_F)
print(n_Fa, n_Sp, n_Lu)
print(n_Sm, n_Me, n_La, n_Ex)

[11  9]
[7 3] [4 6]
[7 0] [1 5] [3 4]
[2 3] [2 5] [3 1] [4 0]


In [6]:
gini_Gender = (wgini(n_M) + wgini(n_F)) / sum(n_01)
gini_CT_Fa = (wgini(n_Fa) + wgini(n_01 - n_Fa)) / sum(n_01)
gini_CT_Sp = (wgini(n_Sp) + wgini(n_01 - n_Sp)) / sum(n_01)
gini_CT_Lu = (wgini(n_Lu) + wgini(n_01 - n_Lu)) / sum(n_01)
gini_SS_Sm = (wgini(n_Sm) + wgini(n_01 - n_Sm)) / sum(n_01)
gini_SS_Me = (wgini(n_Me) + wgini(n_01 - n_Me)) / sum(n_01)
gini_SS_La = (wgini(n_La) + wgini(n_01 - n_La)) / sum(n_01)
gini_SS_Ex = (wgini(n_Ex) + wgini(n_01 - n_Ex)) / sum(n_01)

print('Gender:', gini_Gender)
print('CT Fa:', gini_CT_Fa)
print('CT Sp:', gini_CT_Sp)
print('CT Lu:', gini_CT_Lu)
print('SS Sm:', gini_SS_Sm)
print('SS Me:', gini_SS_Me)
print('SS La:', gini_SS_La)
print('SS Ex:', gini_SS_Ex)

Gender: 0.45
CT Fa: 0.27692307692307694
CT Sp: 0.369047619047619
CT Lu: 0.4791208791208791
SS Sm: 0.48
SS Me: 0.4197802197802198
SS La: 0.475
SS Ex: 0.39375


### Split root by `'Car Type'` being `'Family'` or not (gini index ~= 0.2769)

```
                 ┌ T -> node1
root -> CT == Fa ┤
                 └ F -> node2
```

In [7]:
node1 = root[root[:, 2] == 'Family']
node2 = root[root[:, 2] != 'Family']

## NODE1

In [8]:
print(node1)

[['1' 'M' 'Family' 'Small' 'C0']
 ['2' 'M' 'Family' 'Medium' 'C0']
 ['3' 'M' 'Family' 'Extra Large' 'C0']
 ['5' 'M' 'Family' 'Large' 'C0']
 ['6' 'M' 'Family' 'Extra Large' 'C0']
 ['8' 'F' 'Family' 'Small' 'C0']
 ['9' 'F' 'Family' 'Medium' 'C0']]


### All C0. Node1 is a leaf.

```
node1 -> C0
```

## NODE2

In [9]:
print(node2)

[['4' 'M' 'Sports' 'Extra Large' 'C0']
 ['7' 'M' 'Luxury' 'Extra Large' 'C0']
 ['10' 'F' 'Luxury' 'Large' 'C0']
 ['11' 'F' 'Luxury' 'Large' 'C0']
 ['12' 'M' 'Sports' 'Medium' 'C1']
 ['13' 'M' 'Sports' 'Large' 'C1']
 ['14' 'M' 'Sports' 'Medium' 'C1']
 ['15' 'F' 'Sports' 'Small' 'C1']
 ['16' 'F' 'Luxury' 'Small' 'C1']
 ['17' 'F' 'Luxury' 'Small' 'C1']
 ['18' 'F' 'Sports' 'Medium' 'C1']
 ['19' 'F' 'Luxury' 'Medium' 'C1']
 ['20' 'F' 'Luxury' 'Medium' 'C1']]


In [10]:
n_01 = np.array([0, 0])
n_M, n_F = np.array([0, 0]), np.array([0, 0])
n_Fa, n_Sp, n_Lu = np.array([0, 0]), np.array([0, 0]), np.array([0, 0])
n_Sm, n_Me, n_La, n_Ex = np.array([0, 0]), np.array([0, 0]), np.array([0, 0]), np.array([0, 0])

for i in node2[:, 0]:
    row = data[int(i)]

    n_01[1 - int(row[-1] == 'C0')] += 1

    if row[1] == 'M':
        n_M[1 - int(row[-1] == 'C0')] += 1
    if row[1] == 'F':
        n_F[1 - int(row[-1] == 'C0')] += 1

    if row[2] == 'Family':
        n_Fa[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Sports':
        n_Sp[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Luxury':
        n_Lu[1 - int(row[-1] == 'C0')] += 1

    if row[3] == 'Small':
        n_Sm[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Medium':
        n_Me[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Large':
        n_La[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Extra Large':
        n_Ex[1 - int(row[-1] == 'C0')] += 1

print(n_01)
print(n_M, n_F)
print(n_Fa, n_Sp, n_Lu)
print(n_Sm, n_Me, n_La, n_Ex)

[4 9]
[2 3] [2 6]
[0 0] [1 5] [3 4]
[0 3] [0 5] [2 1] [2 0]


In [11]:
gini_Gender = (wgini(n_M) + wgini(n_F)) / sum(n_01)
gini_CT_Fa = (wgini(n_Fa) + wgini(n_01 - n_Fa)) / sum(n_01)
gini_CT_Sp = (wgini(n_Sp) + wgini(n_01 - n_Sp)) / sum(n_01)
gini_CT_Lu = (wgini(n_Lu) + wgini(n_01 - n_Lu)) / sum(n_01)
gini_SS_Sm = (wgini(n_Sm) + wgini(n_01 - n_Sm)) / sum(n_01)
gini_SS_Me = (wgini(n_Me) + wgini(n_01 - n_Me)) / sum(n_01)
gini_SS_La = (wgini(n_La) + wgini(n_01 - n_La)) / sum(n_01)
gini_SS_Ex = (wgini(n_Ex) + wgini(n_01 - n_Ex)) / sum(n_01)

print('Gender:', gini_Gender)
print('CT Fa:', gini_CT_Fa)
print('CT Sp:', gini_CT_Sp)
print('CT Lu:', gini_CT_Lu)
print('SS Sm:', gini_SS_Sm)
print('SS Me:', gini_SS_Me)
print('SS La:', gini_SS_La)
print('SS Ex:', gini_SS_Ex)

Gender: 0.4153846153846154
CT Fa: nan
CT Sp: 0.39194139194139194
CT Lu: 0.39194139194139194
SS Sm: 0.3692307692307692
SS Me: 0.3076923076923077
SS La: 0.3487179487179486
SS Ex: 0.25174825174825166


/var/folders/1s/h26s0qxx3_b3r_s010t_9dyc0000gn/T/ipykernel_52898/1676990385.py:3: RuntimeWarning: invalid value encountered in long_scalars
  return 1 - (n[0] / s) ** 2 - (n[1] / s) ** 2


### Split node2 by `'Shirt Size'` being `'Extra Large'` or not (gini index ~= 0.2517)

```
                  ┌ T -> node3
node2 -> SS == Ex ┤
                  └ F -> node4
```

In [12]:
node3 = node2[node2[:, 3] == 'Extra Large']
node4 = node2[node2[:, 3] != 'Extra Large']

## NODE3

In [13]:
print(node3)

[['4' 'M' 'Sports' 'Extra Large' 'C0']
 ['7' 'M' 'Luxury' 'Extra Large' 'C0']]


### All C0. Node3 is a leaf.

```
node3 -> C0
```

## NODE4

In [14]:
print(node4)

[['10' 'F' 'Luxury' 'Large' 'C0']
 ['11' 'F' 'Luxury' 'Large' 'C0']
 ['12' 'M' 'Sports' 'Medium' 'C1']
 ['13' 'M' 'Sports' 'Large' 'C1']
 ['14' 'M' 'Sports' 'Medium' 'C1']
 ['15' 'F' 'Sports' 'Small' 'C1']
 ['16' 'F' 'Luxury' 'Small' 'C1']
 ['17' 'F' 'Luxury' 'Small' 'C1']
 ['18' 'F' 'Sports' 'Medium' 'C1']
 ['19' 'F' 'Luxury' 'Medium' 'C1']
 ['20' 'F' 'Luxury' 'Medium' 'C1']]


In [15]:
n_01 = np.array([0, 0])
n_M, n_F = np.array([0, 0]), np.array([0, 0])
n_Fa, n_Sp, n_Lu = np.array([0, 0]), np.array([0, 0]), np.array([0, 0])
n_Sm, n_Me, n_La, n_Ex = np.array([0, 0]), np.array([0, 0]), np.array([0, 0]), np.array([0, 0])

for i in node4[:, 0]:
    row = data[int(i)]

    n_01[1 - int(row[-1] == 'C0')] += 1

    if row[1] == 'M':
        n_M[1 - int(row[-1] == 'C0')] += 1
    if row[1] == 'F':
        n_F[1 - int(row[-1] == 'C0')] += 1

    if row[2] == 'Family':
        n_Fa[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Sports':
        n_Sp[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Luxury':
        n_Lu[1 - int(row[-1] == 'C0')] += 1

    if row[3] == 'Small':
        n_Sm[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Medium':
        n_Me[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Large':
        n_La[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Extra Large':
        n_Ex[1 - int(row[-1] == 'C0')] += 1

print(n_01)
print(n_M, n_F)
print(n_Fa, n_Sp, n_Lu)
print(n_Sm, n_Me, n_La, n_Ex)

[2 9]
[0 3] [2 6]
[0 0] [0 5] [2 4]
[0 3] [0 5] [2 1] [0 0]


In [16]:
gini_Gender = (wgini(n_M) + wgini(n_F)) / sum(n_01)
gini_CT_Fa = (wgini(n_Fa) + wgini(n_01 - n_Fa)) / sum(n_01)
gini_CT_Sp = (wgini(n_Sp) + wgini(n_01 - n_Sp)) / sum(n_01)
gini_CT_Lu = (wgini(n_Lu) + wgini(n_01 - n_Lu)) / sum(n_01)
gini_SS_Sm = (wgini(n_Sm) + wgini(n_01 - n_Sm)) / sum(n_01)
gini_SS_Me = (wgini(n_Me) + wgini(n_01 - n_Me)) / sum(n_01)
gini_SS_La = (wgini(n_La) + wgini(n_01 - n_La)) / sum(n_01)
gini_SS_Ex = (wgini(n_Ex) + wgini(n_01 - n_Ex)) / sum(n_01)

print('Gender:', gini_Gender)
print('CT Fa:', gini_CT_Fa)
print('CT Sp:', gini_CT_Sp)
print('CT Lu:', gini_CT_Lu)
print('SS Sm:', gini_SS_Sm)
print('SS Me:', gini_SS_Me)
print('SS La:', gini_SS_La)
print('SS Ex:', gini_SS_Ex)

Gender: 0.2727272727272727
CT Fa: nan
CT Sp: 0.2424242424242424
CT Lu: 0.2424242424242424
SS Sm: 0.2727272727272727
SS Me: 0.2424242424242424
SS La: 0.12121212121212123
SS Ex: nan


/var/folders/1s/h26s0qxx3_b3r_s010t_9dyc0000gn/T/ipykernel_52898/1676990385.py:3: RuntimeWarning: invalid value encountered in long_scalars
  return 1 - (n[0] / s) ** 2 - (n[1] / s) ** 2


### Split node4 by `'Shirt Size'` being `'Large'` or not (gini index ~= 0.1212)

```
                  ┌ T -> node5
node4 -> SS == La ┤
                  └ F -> node6
```

In [17]:
node5 = node4[node4[:, 3] == 'Large']
node6 = node4[node4[:, 3] != 'Large']

## NODE5

In [18]:
print(node5)

[['10' 'F' 'Luxury' 'Large' 'C0']
 ['11' 'F' 'Luxury' 'Large' 'C0']
 ['13' 'M' 'Sports' 'Large' 'C1']]


In [19]:
n_01 = np.array([0, 0])
n_M, n_F = np.array([0, 0]), np.array([0, 0])
n_Fa, n_Sp, n_Lu = np.array([0, 0]), np.array([0, 0]), np.array([0, 0])
n_Sm, n_Me, n_La, n_Ex = np.array([0, 0]), np.array([0, 0]), np.array([0, 0]), np.array([0, 0])

for i in node5[:, 0]:
    row = data[int(i)]

    n_01[1 - int(row[-1] == 'C0')] += 1

    if row[1] == 'M':
        n_M[1 - int(row[-1] == 'C0')] += 1
    if row[1] == 'F':
        n_F[1 - int(row[-1] == 'C0')] += 1

    if row[2] == 'Family':
        n_Fa[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Sports':
        n_Sp[1 - int(row[-1] == 'C0')] += 1
    if row[2] == 'Luxury':
        n_Lu[1 - int(row[-1] == 'C0')] += 1

    if row[3] == 'Small':
        n_Sm[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Medium':
        n_Me[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Large':
        n_La[1 - int(row[-1] == 'C0')] += 1
    if row[3] == 'Extra Large':
        n_Ex[1 - int(row[-1] == 'C0')] += 1

print(n_01)
print(n_M, n_F)
print(n_Fa, n_Sp, n_Lu)
print(n_Sm, n_Me, n_La, n_Ex)

[2 1]
[0 1] [2 0]
[0 0] [0 1] [2 0]
[0 0] [0 0] [2 1] [0 0]


In [20]:
gini_Gender = (wgini(n_M) + wgini(n_F)) / sum(n_01)
gini_CT_Fa = (wgini(n_Fa) + wgini(n_01 - n_Fa)) / sum(n_01)
gini_CT_Sp = (wgini(n_Sp) + wgini(n_01 - n_Sp)) / sum(n_01)
gini_CT_Lu = (wgini(n_Lu) + wgini(n_01 - n_Lu)) / sum(n_01)
gini_SS_Sm = (wgini(n_Sm) + wgini(n_01 - n_Sm)) / sum(n_01)
gini_SS_Me = (wgini(n_Me) + wgini(n_01 - n_Me)) / sum(n_01)
gini_SS_La = (wgini(n_La) + wgini(n_01 - n_La)) / sum(n_01)
gini_SS_Ex = (wgini(n_Ex) + wgini(n_01 - n_Ex)) / sum(n_01)

print('Gender:', gini_Gender)
print('CT Fa:', gini_CT_Fa)
print('CT Sp:', gini_CT_Sp)
print('CT Lu:', gini_CT_Lu)
print('SS Sm:', gini_SS_Sm)
print('SS Me:', gini_SS_Me)
print('SS La:', gini_SS_La)
print('SS Ex:', gini_SS_Ex)

Gender: 0.0
CT Fa: nan
CT Sp: 0.0
CT Lu: 0.0
SS Sm: nan
SS Me: nan
SS La: nan
SS Ex: nan


/var/folders/1s/h26s0qxx3_b3r_s010t_9dyc0000gn/T/ipykernel_52898/1676990385.py:3: RuntimeWarning: invalid value encountered in long_scalars
  return 1 - (n[0] / s) ** 2 - (n[1] / s) ** 2


### Split node5 by `'Gender'` (gini index ~= 0.0000)

```
                ┌ M -> node7
node5 -> Gender ┤
                └ F -> node8
```

In [21]:
node7 = node5[node5[:, 1] == 'M']
node8 = node5[node5[:, 1] == 'F']

## NODE6

In [22]:
print(node6)

[['12' 'M' 'Sports' 'Medium' 'C1']
 ['14' 'M' 'Sports' 'Medium' 'C1']
 ['15' 'F' 'Sports' 'Small' 'C1']
 ['16' 'F' 'Luxury' 'Small' 'C1']
 ['17' 'F' 'Luxury' 'Small' 'C1']
 ['18' 'F' 'Sports' 'Medium' 'C1']
 ['19' 'F' 'Luxury' 'Medium' 'C1']
 ['20' 'F' 'Luxury' 'Medium' 'C1']]


### All C1. Node6 is a leaf.

```
node6 -> C1
```

## NODE7

In [23]:
print(node7)

[['13' 'M' 'Sports' 'Large' 'C1']]


### All C1. Node7 is a leaf.

```
node7 -> C1
```

## NODE8

In [24]:
print(node8)

[['10' 'F' 'Luxury' 'Large' 'C0']
 ['11' 'F' 'Luxury' 'Large' 'C0']]


### All C0. Node8 is a leaf.

```
node8 -> C0
```

## TREE

```
                 ┌ T -> node1 -> C0
root -> CT == Fa ┤                        ┌ T -> node3 -> C0                              ┌ M -> node7 -> C1
                 └ F -> node2 -> SS == Ex ┤                        ┌ T -> node5 -> Gender ┤
                                          └ F -> node4 -> SS == La ┤                      └ F -> node8 -> C0
                                                                   └ F -> node6 -> C1
```